<center><h1><font size="6">DataDive Event Notebook</font></h1>
<h2>Author: Steven Yan</h2></center>

# <a id='0'>Table of Contents</a>

- <a href='#2'>Importing Packages</a>
- <a href='#2'>Importing Packages</a>
    - ACS Datanase
- <a href='#3'>State-Level Databases: New Yrok</a>
    - ACS Database
    - Geodata
    - FCC Database
    - Ookla Database
    - Joining Dataframes
- <a href='#4'>Illinois Datasets


In [1]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


# <a id='1'>Importing Packages</a>

In [7]:
import pandas as pd 
import numpy as np
! pip install geopandas
import geopandas as gpd 

# Set maximum number of rows and columns that can be viewed
pd.set_option("display.max_columns", 100)
pd.set_option('display.max_rows', 100)


# <a id='2'>Importing Databases</a>

<a href='#0'>Return to Top</a>

Since there seems to be overlap in the terminology, it's best to take a look at how the various databases uses the same nomenclature as the Census Bureau. Generally, state FIPS codes are usually two digits, and county-level FIPS codes are either three or five digits since the state and county FIPS codes can be concatenated to make the five-digit code.


## ACS Database

In [3]:
# Importing ACS database
data = pd.read_csv('/content/drive/MyDrive/Datasets/Broadband/National/acs_2019.csv.gz', compression ='gzip')
data.head()

state  county  tract       geoid  f_broadband  f_computer   f_ba  f_black  \
0      1       1  20100  1001020100        0.740       0.815  0.271    0.113   
1      1       1  20200  1001020200        0.669       0.747  0.171    0.582   
2      1       1  20300  1001020300        0.804       0.879  0.184    0.260   
3      1       1  20400  1001020400        0.856       0.926  0.348    0.070   
4      1       1  20500  1001020500        0.890       0.947  0.424    0.192   

   f_hispanic  log_mhi      mhi  population  households  broadband  computer  \
0       0.013    11.01  60208.0      1993.0       709.0       74.0      81.5   
1       0.015    10.69  43958.0      1959.0       688.0       66.9      74.7   
2       0.066    10.92  55345.0      3507.0      1360.0       80.4      87.9   
3       0.011    11.00  59663.0      3878.0      1675.0       85.6      92.6   
4       0.044    11.10  66108.0     10596.0      4483.0       89.0      94.7   

   black  hispanic    mhi.1    ba  den_computers  n_computer  n_broadband  \
0   11.3       1.3  60208.0  27.1          709.0       578.0        525.0   
1   58.2       1.5  43958.0  17.1          688.0       514.0        460.0   
2   26.0       6.6  55345.0  18.4         1360.0      1196.0       1093.0   
3    7.0       1.1  59663.0  34.8         1675.0      1551.0       1434.0   
4   19.2       4.4  66108.0  42.4         4483.0      4246.0       3991.0   

   den_black  n_black  den_hispanic  n_hispanic  den_ba    n_ba  nhh_computer  \
0     1993.0    226.0        1993.0        26.0  1323.0   359.0           578   
1     1959.0   1141.0        1959.0        30.0  1403.0   240.0           514   
2     3507.0    912.0        3507.0       233.0  2505.0   462.0          1196   
3     3878.0    273.0        3878.0        42.0  2856.0   994.0          1551   
4    10596.0   2038.0       10596.0       469.0  6792.0  2878.0          4246   

   nhh_broadband  nhh_computer_any_internet  nhh_computer_and_dialup  \
0            525                        578                        0   
1            460                        514                        0   
2           1093                       1196                        0   
3           1434                       1551                        7   
4           3991                       4246                        0   

   nhh_computer_and_broadband  nhh_computer_no_internet  nhh_no_computer  \
0                         525                        53              131   
1                         460                        54              174   
2                        1068                       128              164   
3                        1429                       115              124   
4                        3991                       255              237   

   den_age  n_children  n_children_computer  n_children_computer_and_dialup  \
0     1993         414                  414                               0   
1     1750         419                  369                               0   
2     3507         807                  723                               0   
3     3878         712                  712                               0   
4    10418        2678                 2678                               0   

   n_children_computer_and_broadband  n_children_computer_no_internet  \
0                                379                               35   
1                                369                                0   
2                                723                                0   
3                                712                                0   
4                               2554                              124   

   n_children_no_computer  
0                       0  
1                      50  
2                      84  
3                       0  
4                       0

In [4]:
# Get the counts of the values in tract column
data.tract.value_counts()

950100    382
950200    343
100       334
200       331
300       328
         ... 
531302      1
945000      1
441202      1
91017       1
421902      1
Name: tract, Length: 23816, dtype: int64

In [5]:
# Get the counts of the values in geoid column
data.geoid.value_counts()

29189218300    1
2122001200     1
1125010204     1
17167002400    1
22103040801    1
              ..
6085512603     1
34027046298    1
39123050301    1
51059482001    1
13135050418    1
Name: geoid, Length: 73056, dtype: int64

In [6]:
data.columns

Index(['state', 'county', 'tract', 'geoid', 'f_broadband', 'f_computer',
       'f_ba', 'f_black', 'f_hispanic', 'log_mhi', 'mhi', 'population',
       'households', 'broadband', 'computer', 'black', 'hispanic', 'mhi.1',
       'ba', 'den_computers', 'n_computer', 'n_broadband', 'den_black',
       'n_black', 'den_hispanic', 'n_hispanic', 'den_ba', 'n_ba',
       'nhh_computer', 'nhh_broadband', 'nhh_computer_any_internet',
       'nhh_computer_and_dialup', 'nhh_computer_and_broadband',
       'nhh_computer_no_internet', 'nhh_no_computer', 'den_age', 'n_children',
       'n_children_computer', 'n_children_computer_and_dialup',
       'n_children_computer_and_broadband', 'n_children_computer_no_internet',
       'n_children_no_computer'],
      dtype='object')

The columns for the database are:

- `state`
- `county`
- `tract`
- `geoid`
- `f_broadband`, `f_computer`, `f_ba`, `f_black`, `f_hispanic`
- `log_mhi`
- `mhi`
- `population`
- `households`
- `broadband`
- `computer`
- `black`
- `hispanic`
- `mhi.1`
- `ba`
- `den_computers`, `den_age`, `den_black`, `den_hispanic`, `den_ba`
- `n_computer`, `n_broadband`,`n_black`, `n_hispanic`. `n_ba`, `n_children`, `n_children_computer`, `n_children_computer_and_dialup`, `n_children_computer_and_broadband`, `n_children_computer_no_internet`, `n_children_no_computer`
- `nhh_computer`, `nhh_broadband`, `nhh_computer_any_internet`, `nhh_computer_and_dialup`, `nhh_computer_and_broadband`, `nhh_computer_no_internet`, `nhh_no_computer` 

I haven't been able to entirely figure out most of the columns mean.  n is either no or number.  nhh is no household?





# <a id='2'>State-level databases</a>

<a href='#0'>Return to Top</a>

## Geodata


In [7]:
# import shapefile into dataframe
ny_shp = gpd.read_file("/content/drive/MyDrive/Datasets/Broadband/NY/ny_spdf/ny_spdf.shp")
ny_shp.head()

STATEFP COUNTYFP TRACTCE              AFFGEOID        GEOID     NAME LSAD  \
0      36      071  000502  1400000US36071000502  36071000502     5.02   CT   
1      36      103  135208  1400000US36103135208  36103135208  1352.08   CT   
2      36      047  057800  1400000US36047057800  36047057800      578   CT   
3      36      047  058900  1400000US36047058900  36047058900      589   CT   
4      36      055  013204  1400000US36055013204  36055013204   132.04   CT   

        ALAND    AWATER                                           geometry  
0    967431.0  969216.0  POLYGON ((-74.02226 41.49281, -74.02180 41.496...  
1   2287077.0       0.0  POLYGON ((-73.28263 40.83063, -73.28157 40.832...  
2    172233.0       0.0  POLYGON ((-73.95398 40.60140, -73.95304 40.601...  
3    424025.0   38353.0  POLYGON ((-73.94605 40.72926, -73.94419 40.729...  
4  28207247.0   93541.0  POLYGON ((-77.66835 43.02829, -77.66806 43.029...

In [8]:
# get overview of dataframe
ny_shp.info()

<class 'geopandas.geodataframe.GeoDataFrame'>
RangeIndex: 4906 entries, 0 to 4905
Data columns (total 10 columns):
 #   Column    Non-Null Count  Dtype   
---  ------    --------------  -----   
 0   STATEFP   4906 non-null   object  
 1   COUNTYFP  4906 non-null   object  
 2   TRACTCE   4906 non-null   object  
 3   AFFGEOID  4906 non-null   object  
 4   GEOID     4906 non-null   object  
 5   NAME      4906 non-null   object  
 6   LSAD      4906 non-null   object  
 7   ALAND     4906 non-null   float64 
 8   AWATER    4906 non-null   float64 
 9   geometry  4906 non-null   geometry
dtypes: float64(2), geometry(1), object(7)
memory usage: 383.4+ KB


In [9]:
# get number of different county codes
ny_shp.COUNTYFP.nunique()

62

In [10]:
# Get maximum value for county code
ny_shp.COUNTYFP.max()

'123'

In [11]:
# get some information about GEOID
print("Number of unique values: ", ny_shp.GEOID.nunique())
print("Minimum value: ", ny_shp.GEOID.min())
print("Maximum value: ", ny_shp.GEOID.max())

Number of unique values:  4906
Minimum value:  36001000100
Maximum value:  36123150500


In [12]:
# get some information about TRACTCE
print("Number of unique values: ", ny_shp.TRACTCE.nunique())
print("Minimum value: ", ny_shp.TRACTCE.min())
print("Maximum value: ", ny_shp.TRACTCE.max())

Number of unique values:  2697
Minimum value:  000100
Maximum value:  990100


In [13]:
# get value counts of TRACTCE
ny_df = ny_shp.TRACTCE.value_counts().rename_axis('unique_values').reset_index(name='counts')
ny_df.head(10)

unique_values  counts
0        000600      12
1        000700      12
2        000200      12
3        010400      11
4        010200      11
5        013800      11
6        011000      11
7        001600      11
8        000100      11
9        011800      10

In [14]:
ny_000600 = ny_shp[ny_shp['TRACTCE']=='000600']
ny_000600.head()

STATEFP COUNTYFP TRACTCE              AFFGEOID        GEOID NAME LSAD  \
135       36      001  000600  1400000US36001000600  36001000600    6   CT   
830       36      007  000600  1400000US36007000600  36007000600    6   CT   
1286      36      061  000600  1400000US36061000600  36061000600    6   CT   
1359      36      015  000600  1400000US36015000600  36015000600    6   CT   
1573      36      029  000600  1400000US36029000600  36029000600    6   CT   

          ALAND    AWATER                                           geometry  
135    506881.0       0.0  POLYGON ((-73.77924 42.67116, -73.77148 42.666...  
830   1551772.0   12657.0  POLYGON ((-75.89499 42.10717, -75.89437 42.108...  
1286   240406.0  176018.0  POLYGON ((-73.99256 40.71439, -73.99022 40.714...  
1359  1117293.0   39031.0  POLYGON ((-76.82248 42.09642, -76.81913 42.097...  
1573  1163904.0       0.0  POLYGON ((-78.82385 42.83482, -78.82382 42.837...

In [15]:
# sort dataframe by GEOID values
df_1 = ny_shp.sort_values(by=['GEOID'])
df_1.head()

STATEFP COUNTYFP TRACTCE              AFFGEOID        GEOID  NAME LSAD  \
539       36      001  000100  1400000US36001000100  36001000100     1   CT   
2748      36      001  000200  1400000US36001000200  36001000200     2   CT   
3007      36      001  000300  1400000US36001000300  36001000300     3   CT   
3515      36      001  000401  1400000US36001000401  36001000401  4.01   CT   
2751      36      001  000403  1400000US36001000403  36001000403  4.03   CT   

          ALAND    AWATER                                           geometry  
539   2367456.0  245021.0  POLYGON ((-73.74506 42.67228, -73.74374 42.677...  
2748  2065161.0       0.0  POLYGON ((-73.75993 42.65997, -73.76156 42.660...  
3007  5758852.0   55326.0  POLYGON ((-73.83021 42.69418, -73.82785 42.696...  
3515  9065815.0   86327.0  POLYGON ((-73.89421 42.70977, -73.89668 42.711...  
2751  3138700.0       0.0  POLYGON ((-73.82108 42.67857, -73.81916 42.680...

## Synopsis of NY Shapefile:

There are 4906 rows in the shapefile, as well 
as 4906 unique rows of GEOID's. The state FIPS code is 36, and there are 62 county FIPS code in NY.  The AFFGEOID seems to be the GEOID affixed with 1400000US in front of each GEOID.  There are values given for land area, water area, and most importantly the geometry of the tract boundaries.

Originally, I thought that TRACTCEs needed to be concatentated together as they each represented a unique value, but instead they represent a 6-digit tract FIPS code.  

Each GEOID consists of the state FIPS code, its county FIPS code, and then the tract FIPS code, which apparently in itself is not unique and can be repeated, so 2 + 3 + 6 digits in total for the three components or an 11-digit tract FIPS code that is unique for every tract.

## ACS Database

In [16]:
# import database
ny_1 = pd.read_csv('/content/drive/MyDrive/Datasets/Broadband/NY/acs_2019_NY.csv')
ny_1.head()

state  county  tract        geoid  f_broadband  f_computer   f_ba  f_black  \
0     36       1    100  36001000100        0.836       0.896  0.137    0.740   
1     36       1    200  36001000200        0.767       0.903  0.225    0.845   
2     36       1    300  36001000300        0.819       0.881  0.289    0.435   
3     36       1    401  36001000401        0.837       0.857  0.476    0.076   
4     36       1    403  36001000403        0.817       0.861  0.591    0.129   

   f_hispanic  log_mhi      mhi  population  households  broadband  computer  \
0       0.128    10.39  32389.0      2035.0       800.0       83.6      89.6   
1       0.043    10.23  27714.0      4793.0      2213.0       76.7      90.3   
2       0.121    10.72  45272.0      6147.0      2362.0       81.9      88.1   
3       0.046    11.22  74274.0      2362.0       997.0       83.7      85.7   
4       0.058    11.22  74426.0      4253.0      2186.0       81.7      86.1   

   black  hispanic    mhi.1    ba  den_computers  n_computer  n_broadband  \
0   74.0      12.8  32389.0  13.7          800.0       717.0        669.0   
1   84.5       4.3  27714.0  22.5         2213.0      1999.0       1697.0   
2   43.5      12.1  45272.0  28.9         2362.0      2081.0       1935.0   
3    7.6       4.6  74274.0  47.6          997.0       854.0        834.0   
4   12.9       5.8  74426.0  59.1         2186.0      1882.0       1786.0   

   den_black  n_black  den_hispanic  n_hispanic  den_ba    n_ba  nhh_computer  \
0     2035.0   1506.0        2035.0       261.0  1283.0   176.0           717   
1     4793.0   4050.0        4793.0       206.0  2710.0   609.0          1999   
2     6147.0   2674.0        6147.0       742.0  3725.0  1078.0          2081   
3     2362.0    180.0        2362.0       109.0  2169.0  1032.0           854   
4     4253.0    548.0        4253.0       246.0  3220.0  1904.0          1882   

   nhh_broadband  nhh_computer_any_internet  nhh_computer_and_dialup  \
0            669                        717                        0   
1           1697                       1999                        0   
2           1935                       2081                       11   
3            834                        854                        0   
4           1786                       1882                        0   

   nhh_computer_and_broadband  nhh_computer_no_internet  nhh_no_computer  \
0                         669                        48               83   
1                        1685                       314              214   
2                        1917                       153              281   
3                         826                        28              143   
4                        1733                       149              304   

   den_age  n_children  n_children_computer  n_children_computer_and_dialup  \
0     2022         532                  520                               0   
1     4711        1274                 1221                               0   
2     5925        1729                 1717                               0   
3     1790         171                  171                               0   
4     4240         462                  462                               0   

   n_children_computer_and_broadband  n_children_computer_no_internet  \
0                                511                                9   
1                               1138                               83   
2                               1630                               87   
3                                171                                0   
4                                462                                0   

   n_children_no_computer  state_lkp  
0                      12         36  
1                      53         36  
2                      12         36  
3                       0         36  
4                       0         36

In [17]:
ny_1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4918 entries, 0 to 4917
Data columns (total 43 columns):
 #   Column                             Non-Null Count  Dtype  
---  ------                             --------------  -----  
 0   state                              4918 non-null   int64  
 1   county                             4918 non-null   int64  
 2   tract                              4918 non-null   int64  
 3   geoid                              4918 non-null   int64  
 4   f_broadband                        4918 non-null   float64
 5   f_computer                         4918 non-null   float64
 6   f_ba                               4918 non-null   float64
 7   f_black                            4918 non-null   float64
 8   f_hispanic                         4918 non-null   float64
 9   log_mhi                            4785 non-null   float64
 10  mhi                                4918 non-null   float64
 11  population                         4918 non-null   float

In [18]:
# check for any null values
ny_1.isna().sum()

state                                  0
county                                 0
tract                                  0
geoid                                  0
f_broadband                            0
f_computer                             0
f_ba                                   0
f_black                                0
f_hispanic                             0
log_mhi                              133
mhi                                    0
population                             0
households                             0
broadband                              0
computer                               0
black                                  0
hispanic                               0
mhi.1                                  0
ba                                     0
den_computers                          0
n_computer                             0
n_broadband                            0
den_black                              0
n_black                                0
den_hispanic    

In [19]:
# get some information about tract
print("Number of unique values: ", ny_1.tract.nunique())
print("Minimum value: ", ny_1.tract.min())
print("Maximum value: ", ny_1.tract.max())

Number of unique values:  2701
Minimum value:  100
Maximum value:  990400


In [20]:
# get some information about geoid
print("Number of unique values: ", ny_1.geoid.nunique())
print("Minimum value: ", ny_1.geoid.min())
print("Maximum value: ", ny_1.geoid.max())


Number of unique values:  4918
Minimum value:  36001000100
Maximum value:  36123150500


In [21]:
ny_1['geoid'] = ny_1['geoid'].apply(lambda x: str(x))

In [22]:
joined_df = df_1.merge(ny_1, how='left', left_on='GEOID', right_on='geoid')
joined_df.head()

STATEFP COUNTYFP TRACTCE              AFFGEOID        GEOID  NAME LSAD  \
0      36      001  000100  1400000US36001000100  36001000100     1   CT   
1      36      001  000200  1400000US36001000200  36001000200     2   CT   
2      36      001  000300  1400000US36001000300  36001000300     3   CT   
3      36      001  000401  1400000US36001000401  36001000401  4.01   CT   
4      36      001  000403  1400000US36001000403  36001000403  4.03   CT   

       ALAND    AWATER                                           geometry  \
0  2367456.0  245021.0  POLYGON ((-73.74506 42.67228, -73.74374 42.677...   
1  2065161.0       0.0  POLYGON ((-73.75993 42.65997, -73.76156 42.660...   
2  5758852.0   55326.0  POLYGON ((-73.83021 42.69418, -73.82785 42.696...   
3  9065815.0   86327.0  POLYGON ((-73.89421 42.70977, -73.89668 42.711...   
4  3138700.0       0.0  POLYGON ((-73.82108 42.67857, -73.81916 42.680...   

   state  county  tract        geoid  f_broadband  f_computer   f_ba  f_black  \
0     36       1    100  36001000100        0.836       0.896  0.137    0.740   
1     36       1    200  36001000200        0.767       0.903  0.225    0.845   
2     36       1    300  36001000300        0.819       0.881  0.289    0.435   
3     36       1    401  36001000401        0.837       0.857  0.476    0.076   
4     36       1    403  36001000403        0.817       0.861  0.591    0.129   

   f_hispanic  log_mhi      mhi  population  households  broadband  computer  \
0       0.128    10.39  32389.0      2035.0       800.0       83.6      89.6   
1       0.043    10.23  27714.0      4793.0      2213.0       76.7      90.3   
2       0.121    10.72  45272.0      6147.0      2362.0       81.9      88.1   
3       0.046    11.22  74274.0      2362.0       997.0       83.7      85.7   
4       0.058    11.22  74426.0      4253.0      2186.0       81.7      86.1   

   black  hispanic    mhi.1    ba  den_computers  n_computer  n_broadband  \
0   74.0      12.8  32389.0  13.7          800.0       717.0        669.0   
1   84.5       4.3  27714.0  22.5         2213.0      1999.0       1697.0   
2   43.5      12.1  45272.0  28.9         2362.0      2081.0       1935.0   
3    7.6       4.6  74274.0  47.6          997.0       854.0        834.0   
4   12.9       5.8  74426.0  59.1         2186.0      1882.0       1786.0   

   den_black  n_black  den_hispanic  n_hispanic  den_ba    n_ba  nhh_computer  \
0     2035.0   1506.0        2035.0       261.0  1283.0   176.0           717   
1     4793.0   4050.0        4793.0       206.0  2710.0   609.0          1999   
2     6147.0   2674.0        6147.0       742.0  3725.0  1078.0          2081   
3     2362.0    180.0        2362.0       109.0  2169.0  1032.0           854   
4     4253.0    548.0        4253.0       246.0  3220.0  1904.0          1882   

   nhh_broadband  nhh_computer_any_internet  nhh_computer_and_dialup  \
0            669                        717                        0   
1           1697                       1999                        0   
2           1935                       2081                       11   
3            834                        854                        0   
4           1786                       1882                        0   

   nhh_computer_and_broadband  nhh_computer_no_internet  nhh_no_computer  \
0                         669                        48               83   
1                        1685                       314              214   
2                        1917                       153              281   
3                         826                        28              143   
4                        1733                       149              304   

   den_age  n_children  n_children_computer  n_children_computer_and_dialup  \
0     2022         532                  520                               0   
1     4711        1274                 1221                               0   
2     5925        1729               

In [23]:
joined_df.drop(columns=['STATEFP', 'COUNTYFP', 'TRACTCE', 'AFFGEOID', 'NAME', 'LSAD', 'ALAND', 'AWATER', 'state', 'county', 'tract', 'geoid', 'state_lkp'], inplace=True)

In [24]:
joined_df.head()

GEOID                                           geometry  \
0  36001000100  POLYGON ((-73.74506 42.67228, -73.74374 42.677...   
1  36001000200  POLYGON ((-73.75993 42.65997, -73.76156 42.660...   
2  36001000300  POLYGON ((-73.83021 42.69418, -73.82785 42.696...   
3  36001000401  POLYGON ((-73.89421 42.70977, -73.89668 42.711...   
4  36001000403  POLYGON ((-73.82108 42.67857, -73.81916 42.680...   

   f_broadband  f_computer   f_ba  f_black  f_hispanic  log_mhi      mhi  \
0        0.836       0.896  0.137    0.740       0.128    10.39  32389.0   
1        0.767       0.903  0.225    0.845       0.043    10.23  27714.0   
2        0.819       0.881  0.289    0.435       0.121    10.72  45272.0   
3        0.837       0.857  0.476    0.076       0.046    11.22  74274.0   
4        0.817       0.861  0.591    0.129       0.058    11.22  74426.0   

   population  households  broadband  computer  black  hispanic    mhi.1  \
0      2035.0       800.0       83.6      89.6   74.0      12.8  32389.0   
1      4793.0      2213.0       76.7      90.3   84.5       4.3  27714.0   
2      6147.0      2362.0       81.9      88.1   43.5      12.1  45272.0   
3      2362.0       997.0       83.7      85.7    7.6       4.6  74274.0   
4      4253.0      2186.0       81.7      86.1   12.9       5.8  74426.0   

     ba  den_computers  n_computer  n_broadband  den_black  n_black  \
0  13.7          800.0       717.0        669.0     2035.0   1506.0   
1  22.5         2213.0      1999.0       1697.0     4793.0   4050.0   
2  28.9         2362.0      2081.0       1935.0     6147.0   2674.0   
3  47.6          997.0       854.0        834.0     2362.0    180.0   
4  59.1         2186.0      1882.0       1786.0     4253.0    548.0   

   den_hispanic  n_hispanic  den_ba    n_ba  nhh_computer  nhh_broadband  \
0        2035.0       261.0  1283.0   176.0           717            669   
1        4793.0       206.0  2710.0   609.0          1999           1697   
2        6147.0       742.0  3725.0  1078.0          2081           1935   
3        2362.0       109.0  2169.0  1032.0           854            834   
4        4253.0       246.0  3220.0  1904.0          1882           1786   

   nhh_computer_any_internet  nhh_computer_and_dialup  \
0                        717                        0   
1                       1999                        0   
2                       2081                       11   
3                        854                        0   
4                       1882                        0   

   nhh_computer_and_broadband  nhh_computer_no_internet  nhh_no_computer  \
0                         669                        48               83   
1                        1685                       314              214   
2                        1917                       153              281   
3                         826                        28              143   
4                        1733                       149              304   

   den_age  n_children  n_children_computer  n_children_computer_and_dialup  \
0     2022         532                  520                               0   
1     4711        1274                 1221                               0   
2     5925        1729                 1717                               0   
3     1790         171                  171                               0   
4     4240         462                  462                               0   

   n_children_computer_and_broadband  n_children_computer_no_internet  \
0                                511                                9   
1                               1138                               83   
2                               1630                               87   
3                                171                                0   
4                                462                                0   

   n_children_no_computer  
0                      12  
1                      53 

In [25]:
type(joined_df)

geopandas.geodataframe.GeoDataFrame

In [26]:
joined_df.info()

<class 'geopandas.geodataframe.GeoDataFrame'>
Int64Index: 4906 entries, 0 to 4905
Data columns (total 40 columns):
 #   Column                             Non-Null Count  Dtype   
---  ------                             --------------  -----   
 0   GEOID                              4906 non-null   object  
 1   geometry                           4906 non-null   geometry
 2   f_broadband                        4906 non-null   float64 
 3   f_computer                         4906 non-null   float64 
 4   f_ba                               4906 non-null   float64 
 5   f_black                            4906 non-null   float64 
 6   f_hispanic                         4906 non-null   float64 
 7   log_mhi                            4785 non-null   float64 
 8   mhi                                4906 non-null   float64 
 9   population                         4906 non-null   float64 
 10  households                         4906 non-null   float64 
 11  broadband                          

## FCC Database

The columns are as follows:

- `tract`: 11-digit tract FIPS code
- `max_dn`: average maximum advertised downstream speed offered by provider in census tract by block
- `max_up`: average maximum advertised upstream speed offered by provider in census tract by block
- `dn10`:	average count of providers offering advertised downstream greater than 10 by block
- `dn100`: average count of providers offering advertised downstream greater than 100 by block
- `dn250`: average count of providers offering advertised downstream greater than 250 by block
- `fiber_100u`: average count of providers offiering fiber by block


In [27]:
# importing FCC database
ny_2 = pd.read_csv('/content/drive/MyDrive/Datasets/Broadband/NY/fcc_477_census_tract_NY.csv')
ny_2.head()

tract      max_dn     max_up      dn10     dn100     dn250  \
0  36001000100  711.250000  50.722222  1.875000  0.777778  0.777778   
1  36001000200  940.000000  35.000000  2.529412  1.011765  1.000000   
2  36001000300  928.040541  98.195946  2.290541  1.060811  1.060811   
3  36001000401  928.975904  65.168675  2.156627  1.024096  1.024096   
4  36001000403  940.000000  43.203883  2.310680  1.009709  1.009709   

   fiber_100u  state  
0    0.027778     36  
1    0.000000     36  
2    0.074324     36  
3    0.036145     36  
4    0.009709     36

In [28]:
# check for any null values in dataframe
ny_2.isna().sum()

tract         0
max_dn        0
max_up        0
dn10          0
dn100         0
dn250         0
fiber_100u    0
state         0
dtype: int64

In [29]:
# show value counts of tract column
ny_2.tract.value_counts()

36029003800    1
36093020800    1
36029004402    1
36027110003    1
36029016700    1
              ..
36029011500    1
36047082800    1
36093021200    1
36005008700    1
36019101700    1
Name: tract, Length: 4900, dtype: int64

In [30]:
# find unique values of tract column
ny_2.tract.nunique()

4900

In [31]:
# change tract column to string type for joining
ny_2['tract'] = ny_2['tract'].apply(lambda x: str(x))
type(ny_2['tract'][0])

str

In [32]:
# merge two database on the unique 11-digit FIPS code
df_2 = joined_df.merge(ny_2, how='left', left_on='GEOID', right_on='tract')
df_2.head()

GEOID                                           geometry  \
0  36001000100  POLYGON ((-73.74506 42.67228, -73.74374 42.677...   
1  36001000200  POLYGON ((-73.75993 42.65997, -73.76156 42.660...   
2  36001000300  POLYGON ((-73.83021 42.69418, -73.82785 42.696...   
3  36001000401  POLYGON ((-73.89421 42.70977, -73.89668 42.711...   
4  36001000403  POLYGON ((-73.82108 42.67857, -73.81916 42.680...   

   f_broadband  f_computer   f_ba  f_black  f_hispanic  log_mhi      mhi  \
0        0.836       0.896  0.137    0.740       0.128    10.39  32389.0   
1        0.767       0.903  0.225    0.845       0.043    10.23  27714.0   
2        0.819       0.881  0.289    0.435       0.121    10.72  45272.0   
3        0.837       0.857  0.476    0.076       0.046    11.22  74274.0   
4        0.817       0.861  0.591    0.129       0.058    11.22  74426.0   

   population  households  broadband  computer  black  hispanic    mhi.1  \
0      2035.0       800.0       83.6      89.6   74.0      12.8  32389.0   
1      4793.0      2213.0       76.7      90.3   84.5       4.3  27714.0   
2      6147.0      2362.0       81.9      88.1   43.5      12.1  45272.0   
3      2362.0       997.0       83.7      85.7    7.6       4.6  74274.0   
4      4253.0      2186.0       81.7      86.1   12.9       5.8  74426.0   

     ba  den_computers  n_computer  n_broadband  den_black  n_black  \
0  13.7          800.0       717.0        669.0     2035.0   1506.0   
1  22.5         2213.0      1999.0       1697.0     4793.0   4050.0   
2  28.9         2362.0      2081.0       1935.0     6147.0   2674.0   
3  47.6          997.0       854.0        834.0     2362.0    180.0   
4  59.1         2186.0      1882.0       1786.0     4253.0    548.0   

   den_hispanic  n_hispanic  den_ba    n_ba  nhh_computer  nhh_broadband  \
0        2035.0       261.0  1283.0   176.0           717            669   
1        4793.0       206.0  2710.0   609.0          1999           1697   
2        6147.0       742.0  3725.0  1078.0          2081           1935   
3        2362.0       109.0  2169.0  1032.0           854            834   
4        4253.0       246.0  3220.0  1904.0          1882           1786   

   nhh_computer_any_internet  nhh_computer_and_dialup  \
0                        717                        0   
1                       1999                        0   
2                       2081                       11   
3                        854                        0   
4                       1882                        0   

   nhh_computer_and_broadband  nhh_computer_no_internet  nhh_no_computer  \
0                         669                        48               83   
1                        1685                       314              214   
2                        1917                       153              281   
3                         826                        28              143   
4                        1733                       149              304   

   den_age  n_children  n_children_computer  n_children_computer_and_dialup  \
0     2022         532                  520                               0   
1     4711        1274                 1221                               0   
2     5925        1729                 1717                               0   
3     1790         171                  171                               0   
4     4240         462                  462                               0   

   n_children_computer_and_broadband  n_children_computer_no_internet  \
0                                511                                9   
1                               1138                               83   
2                               1630                               87   
3                                171                                0   
4                                462                                0   

   n_children_no_computer        tract      max_dn     max_up      dn10  \
0      

In [33]:
df_2.drop(columns=['tract', 'state'], inplace=True)

In [34]:
df_2.info()

<class 'geopandas.geodataframe.GeoDataFrame'>
Int64Index: 4906 entries, 0 to 4905
Data columns (total 46 columns):
 #   Column                             Non-Null Count  Dtype   
---  ------                             --------------  -----   
 0   GEOID                              4906 non-null   object  
 1   geometry                           4906 non-null   geometry
 2   f_broadband                        4906 non-null   float64 
 3   f_computer                         4906 non-null   float64 
 4   f_ba                               4906 non-null   float64 
 5   f_black                            4906 non-null   float64 
 6   f_hispanic                         4906 non-null   float64 
 7   log_mhi                            4785 non-null   float64 
 8   mhi                                4906 non-null   float64 
 9   population                         4906 non-null   float64 
 10  households                         4906 non-null   float64 
 11  broadband                          

In [35]:
type(df_2)

geopandas.geodataframe.GeoDataFrame

## Synopsis of FCC Database

FCC has 4900 rows of data, and the shapefile had 4906 rows of each of the tracts, and every row has a unique tract value. Here `tract` is synonymous with `GEOID`. It would seem we are missing data for a few GEOIDs, but that remains to be seen how our analysis will be affected.

## Ookla Database

The column keys are:
- `avg_d_kbps`, `int`, average download speed of all tests performed in the tile, represented in kilobits per second.
- `avg_u_kbps`, `int`, 	average upload speed of all tests performed in the tile, represented in kilobits per second.
- `avg_lat_ms`, `int`, average latency of all tests performed in the tile, represented in milliseconds
- `tests`, `int`, number of tests taken in the tile.
- `devices`, `int`, number of unique devices contributing tests in the tile.
- `quadkey`: quadkey representing the tile.



In [48]:
# import Ookla database
ny_3 = pd.read_csv('/content/drive/MyDrive/Datasets/Broadband/NY/ookla_combined_ny.csv')
ny_3.head()

quadkey  avg_d_kbps  avg_u_kbps  avg_lat_ms  tests  devices   type  \
0  302332203102320      208526        7394          14     50        2  fixed   
1  302231312100203      153742       97895           6    134       28  fixed   
2  320101103112201      202082       55540          15    222       56  fixed   
3  302322000030131      176743       14699          20     18        3  fixed   
4  302233111212211      370517      311427           4      2        1  fixed   

  quarter  year  STATEFP  COUNTYFP  TRACTCE              AFFGEOID  \
0      Q1  2021       36       103   170202  1400000US36103170202   
1      Q1  2021       36        63    24502  1400000US36063024502   
2      Q1  2021       36        47    92400  1400000US36047092400   
3      Q1  2021       36        29    14702  1400000US36029014702   
4      Q1  2021       36        29     7302  1400000US36029007302   

         GEOID     NAME LSAD     ALAND     AWATER  \
0  36103170202  1702.02   CT  41367468  100616534   
1  36063024502   245.02   CT  77726398   10786358   
2  36047092400   924.00   CT    251687          0   
3  36029014702   147.02   CT  31997799     308205   
4  36029007302    73.02   CT  36949030    5791136   

                                                tile  
0  list(c(-72.57568359375, -72.5701904296875, -72...  
1  list(c(-79.2718505859375, -79.266357421875, -7...  
2  list(c(-73.9105224609375, -73.905029296875, -7...  
3  list(c(-78.6236572265625, -78.6181640625, -78....  
4  list(c(-78.9971923828125, -78.99169921875, -78...

In [49]:
# get general attributes of dataframe
ny_3.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 444771 entries, 0 to 444770
Data columns (total 19 columns):
 #   Column      Non-Null Count   Dtype  
---  ------      --------------   -----  
 0   quadkey     444771 non-null  int64  
 1   avg_d_kbps  444771 non-null  int64  
 2   avg_u_kbps  444771 non-null  int64  
 3   avg_lat_ms  444771 non-null  int64  
 4   tests       444771 non-null  int64  
 5   devices     444771 non-null  int64  
 6   type        444771 non-null  object 
 7   quarter     444771 non-null  object 
 8   year        444771 non-null  int64  
 9   STATEFP     444771 non-null  int64  
 10  COUNTYFP    444771 non-null  int64  
 11  TRACTCE     444771 non-null  int64  
 12  AFFGEOID    444771 non-null  object 
 13  GEOID       444771 non-null  int64  
 14  NAME        444771 non-null  float64
 15  LSAD        444771 non-null  object 
 16  ALAND       444771 non-null  int64  
 17  AWATER      444771 non-null  int64  
 18  tile        444771 non-null  object 
dtypes:

In [50]:
# get unique values of GEOID
ny_3.GEOID.nunique()

3891

The number of rows are significantly higher than the other databases, so there are multiple instances of each tract area.  There are also only 3891 different tract areas represented, which is more than 1000 less than the total off 4906 in the shapefiles.

In [53]:
# check for nnull valaues
ny_3.isna().sum()

quadkey       0
avg_d_kbps    0
avg_u_kbps    0
avg_lat_ms    0
tests         0
devices       0
type          0
quarter       0
year          0
STATEFP       0
COUNTYFP      0
TRACTCE       0
AFFGEOID      0
GEOID         0
NAME          0
LSAD          0
ALAND         0
AWATER        0
tile          0
dtype: int64

In [54]:
# get unique values of tile column
ny_3.tile.nunique()

234446

In [55]:
# get unique values of quadkeys column
ny_3.quadkey.nunique()

234446

I need to incorporate Group 1A's code for weighted averages, but I aggregated the data by taking the average or sum depending on the column.

In [56]:
import numpy as np
wm = lambda x: np.average(x, weights=ny_3.loc[x.index, 'tests'])

df_3 = ny_3.groupby(['GEOID','type'], as_index=False).agg(
            d_kbps_wm = ('avg_d_kbps', wm),
            u_kbps_wm = ('avg_u_kbps', wm)
)


In [57]:
fixed = df_3[df_3.type == 'fixed']
mobile = df_3[df_3.type == 'mobile']
fixed.head()

GEOID   type      d_kbps_wm     u_kbps_wm
0  36001000100  fixed  163889.809917  21215.471074
2  36001000200  fixed  187803.795181  12906.168675
4  36001000300  fixed  168756.639175  28695.003436
6  36001000401  fixed  112107.778523  17913.073826
8  36001000403  fixed  169190.106667  17986.471111

In [60]:
df_3a = mobile.merge(fixed, on='GEOID')
df_3a.head()

GEOID  type_x   d_kbps_wm_x   u_kbps_wm_x type_y    d_kbps_wm_y  \
0  36001000100  mobile  65457.218750  15323.062500  fixed  163889.809917   
1  36001000200  mobile  21035.480000   9172.693333  fixed  187803.795181   
2  36001000300  mobile  80416.212121  18435.893939  fixed  168756.639175   
3  36001000401  mobile  71913.388235  17522.258824  fixed  112107.778523   
4  36001000403  mobile  95159.347368  24303.431579  fixed  169190.106667   

    u_kbps_wm_y  
0  21215.471074  
1  12906.168675  
2  28695.003436  
3  17913.073826  
4  17986.471111

In [61]:
# use groupby to aggregate rows to turn into dataframe
df_3b = ny_3.groupby('GEOID', as_index=False).agg({'tests': 'sum', 'devices': 'sum'})
df_3b.head()

GEOID  tests  devices
0  36001000100    153       68
1  36001000200    158       67
2  36001000300    357      135
3  36001000401    319      108
4  36001000403    320      163

In [ ]:
df_3b = ny_3.groupby('GEOID', as_index=False).agg({'tests': 'sum', 'devices': 'sum'})
df_4 = df_3a.merge(df_3b, on="GEOID")
df_4['GEOID'] = df_4['GEOID'].apply(lambda x: str(x))
df_tot = df_2.merge(df_4, how='left', on='GEOID')

In [62]:
df_4 = df_3a.merge(df_3b, on="GEOID")
df_4.head()

GEOID  type_x   d_kbps_wm_x   u_kbps_wm_x type_y    d_kbps_wm_y  \
0  36001000100  mobile  65457.218750  15323.062500  fixed  163889.809917   
1  36001000200  mobile  21035.480000   9172.693333  fixed  187803.795181   
2  36001000300  mobile  80416.212121  18435.893939  fixed  168756.639175   
3  36001000401  mobile  71913.388235  17522.258824  fixed  112107.778523   
4  36001000403  mobile  95159.347368  24303.431579  fixed  169190.106667   

    u_kbps_wm_y  tests  devices  
0  21215.471074    153       68  
1  12906.168675    158       67  
2  28695.003436    357      135  
3  17913.073826    319      108  
4  17986.471111    320      163

In [63]:
df_4['GEOID'] = df_4['GEOID'].apply(lambda x: str(x))

Need to look into joining tiles to the counties: 

```
tiles_in_ky_counties = gpd.sjoin(tiles, ky_counties, how='inner', op='intersects')
```

May account for the missing areas in this dataset.  The Ookla dataset we were provided is not a geodataframe, and sjoin only works on two geodataframes.  I managed to download a geo version of the ookla dataset, but have not incorporated yet into notebook.

In [64]:
df_tot = df_2.merge(df_4, how='left', on='GEOID')
df_tot.head()


GEOID                                           geometry  \
0  36001000100  POLYGON ((-73.74506 42.67228, -73.74374 42.677...   
1  36001000200  POLYGON ((-73.75993 42.65997, -73.76156 42.660...   
2  36001000300  POLYGON ((-73.83021 42.69418, -73.82785 42.696...   
3  36001000401  POLYGON ((-73.89421 42.70977, -73.89668 42.711...   
4  36001000403  POLYGON ((-73.82108 42.67857, -73.81916 42.680...   

   f_broadband  f_computer   f_ba  f_black  f_hispanic  log_mhi      mhi  \
0        0.836       0.896  0.137    0.740       0.128    10.39  32389.0   
1        0.767       0.903  0.225    0.845       0.043    10.23  27714.0   
2        0.819       0.881  0.289    0.435       0.121    10.72  45272.0   
3        0.837       0.857  0.476    0.076       0.046    11.22  74274.0   
4        0.817       0.861  0.591    0.129       0.058    11.22  74426.0   

   population  households  broadband  computer  black  hispanic    mhi.1  \
0      2035.0       800.0       83.6      89.6   74.0      12.8  32389.0   
1      4793.0      2213.0       76.7      90.3   84.5       4.3  27714.0   
2      6147.0      2362.0       81.9      88.1   43.5      12.1  45272.0   
3      2362.0       997.0       83.7      85.7    7.6       4.6  74274.0   
4      4253.0      2186.0       81.7      86.1   12.9       5.8  74426.0   

     ba  den_computers  n_computer  n_broadband  den_black  n_black  \
0  13.7          800.0       717.0        669.0     2035.0   1506.0   
1  22.5         2213.0      1999.0       1697.0     4793.0   4050.0   
2  28.9         2362.0      2081.0       1935.0     6147.0   2674.0   
3  47.6          997.0       854.0        834.0     2362.0    180.0   
4  59.1         2186.0      1882.0       1786.0     4253.0    548.0   

   den_hispanic  n_hispanic  den_ba    n_ba  nhh_computer  nhh_broadband  \
0        2035.0       261.0  1283.0   176.0           717            669   
1        4793.0       206.0  2710.0   609.0          1999           1697   
2        6147.0       742.0  3725.0  1078.0          2081           1935   
3        2362.0       109.0  2169.0  1032.0           854            834   
4        4253.0       246.0  3220.0  1904.0          1882           1786   

   nhh_computer_any_internet  nhh_computer_and_dialup  \
0                        717                        0   
1                       1999                        0   
2                       2081                       11   
3                        854                        0   
4                       1882                        0   

   nhh_computer_and_broadband  nhh_computer_no_internet  nhh_no_computer  \
0                         669                        48               83   
1                        1685                       314              214   
2                        1917                       153              281   
3                         826                        28              143   
4                        1733                       149              304   

   den_age  n_children  n_children_computer  n_children_computer_and_dialup  \
0     2022         532                  520                               0   
1     4711        1274                 1221                               0   
2     5925        1729                 1717                               0   
3     1790         171                  171                               0   
4     4240         462                  462                               0   

   n_children_computer_and_broadband  n_children_computer_no_internet  \
0                                511                                9   
1                               1138                               83   
2                               1630                               87   
3                                171                                0   
4                                462                                0   

   n_children_no_computer      max_dn     max_up      dn10     dn100  \
0         

In [65]:
df_tot.info()

<class 'geopandas.geodataframe.GeoDataFrame'>
Int64Index: 4906 entries, 0 to 4905
Data columns (total 54 columns):
 #   Column                             Non-Null Count  Dtype   
---  ------                             --------------  -----   
 0   GEOID                              4906 non-null   object  
 1   geometry                           4906 non-null   geometry
 2   f_broadband                        4906 non-null   float64 
 3   f_computer                         4906 non-null   float64 
 4   f_ba                               4906 non-null   float64 
 5   f_black                            4906 non-null   float64 
 6   f_hispanic                         4906 non-null   float64 
 7   log_mhi                            4785 non-null   float64 
 8   mhi                                4906 non-null   float64 
 9   population                         4906 non-null   float64 
 10  households                         4906 non-null   float64 
 11  broadband                          

In [66]:
df_tot.isna().sum()

GEOID                                   0
geometry                                0
f_broadband                             0
f_computer                              0
f_ba                                    0
f_black                                 0
f_hispanic                              0
log_mhi                               121
mhi                                     0
population                              0
households                              0
broadband                               0
computer                                0
black                                   0
hispanic                                0
mhi.1                                   0
ba                                      0
den_computers                           0
n_computer                              0
n_broadband                             0
den_black                               0
n_black                                 0
den_hispanic                            0
n_hispanic                        

In [67]:
df_tot.to_csv("df_total.csv")

# Illinois Datasets

In [10]:
shapefile = "/content/drive/MyDrive/Datasets/Broadband/IL/il_spdf/il_spdf.shp"
acsfile = '/content/drive/MyDrive/Datasets/Broadband/IL/acs_2019_IL.csv'
fccfile = '/content/drive/MyDrive/Datasets/Broadband/IL/fcc_477_census_tract_IL.csv'
ooklafile = '/content/drive/MyDrive/Datasets/Broadband/IL/ookla_combined_il.csv'

In [11]:
shpfile = gpd.read_file(shapefile)
shpfile = shpfile.sort_values(by=['GEOID'])
shpfile.head()

STATEFP COUNTYFP TRACTCE              AFFGEOID        GEOID  NAME LSAD  \
1992      17      001  000100  1400000US17001000100  17001000100     1   CT   
1980      17      001  000201  1400000US17001000201  17001000201  2.01   CT   
22        17      001  000202  1400000US17001000202  17001000202  2.02   CT   
1117      17      001  000400  1400000US17001000400  17001000400     4   CT   
784       17      001  000500  1400000US17001000500  17001000500     5   CT   

           ALAND    AWATER                                           geometry  
1992  12937185.0   22042.0  POLYGON ((-91.37766 39.94160, -91.37759 39.946...  
1980   3206335.0       0.0  POLYGON ((-91.39646 39.95621, -91.39631 39.965...  
22     1514956.0       0.0  POLYGON ((-91.39370 39.94678, -91.37759 39.946...  
1117   4458010.0  211712.0  POLYGON ((-91.42005 39.95081, -91.41917 39.951...  
784    1262517.0       0.0  POLYGON ((-91.40340 39.95048, -91.39655 39.950...

In [12]:
data = pd.read_csv(acsfile)
data = data.rename(columns={'geoid':'GEOID'})
data['GEOID'] = data['GEOID'].apply(lambda x: str(x))
data.drop(columns=['state', 'county', 'tract', 'state_lkp'], inplace=True)
data.head()

GEOID  f_broadband  f_computer   f_ba  f_black  f_hispanic  log_mhi  \
0  17001000100        0.813       0.861  0.246    0.068       0.000    10.78   
1  17001000201        0.879       0.940  0.266    0.078       0.016    10.71   
2  17001000202        0.863       0.933  0.221    0.080       0.008    10.83   
3  17001000400        0.691       0.780  0.094    0.182       0.006    10.31   
4  17001000500        0.665       0.787  0.113    0.133       0.035    10.62   

       mhi  population  households  broadband  computer  black  hispanic  \
0  48088.0      4761.0      2202.0       81.3      86.1    6.8       0.0   
1  44855.0      1985.0       889.0       87.9      94.0    7.8       1.6   
2  50375.0      2376.0       970.0       86.3      93.3    8.0       0.8   
3  30164.0      3422.0      1316.0       69.1      78.0   18.2       0.6   
4  41008.0      2175.0       813.0       66.5      78.7   13.3       3.5   

     mhi.1    ba  den_computers  n_computer  n_broadband  den_black  n_black  \
0  48088.0  24.6         2202.0      1896.0       1791.0     4761.0    324.0   
1  44855.0  26.6          889.0       836.0        781.0     1985.0    155.0   
2  50375.0  22.1          970.0       905.0        837.0     2376.0    191.0   
3  30164.0   9.4         1316.0      1027.0        909.0     3422.0    624.0   
4  41008.0  11.3          813.0       640.0        541.0     2175.0    290.0   

   den_hispanic  n_hispanic  den_ba   n_ba  nhh_computer  nhh_broadband  \
0        4761.0         0.0  3381.0  831.0          1896           1791   
1        1985.0        32.0  1400.0  373.0           836            781   
2        2376.0        19.0  1527.0  338.0           905            837   
3        3422.0        20.0  2536.0  238.0          1027            909   
4        2175.0        77.0  1455.0  165.0           640            541   

   nhh_computer_any_internet  nhh_computer_and_dialup  \
0                       1896                        0   
1                        836                        5   
2                        905                        7   
3                       1027                        9   
4                        640                        0   

   nhh_computer_and_broadband  nhh_computer_no_internet  nhh_no_computer  \
0                        1791                       105              306   
1                         781                        50               53   
2                         829                        69               65   
3                         909                       109              289   
4                         536                       104              173   

   den_age  n_children  n_children_computer  n_children_computer_and_dialup  \
0     4741        1044                 1044                               0   
1     1961         383                  383                               0   
2     2336         521                  521                               0   
3     3147         645                  637                               0   
4     1878         475                  458                               0   

   n_children_computer_and_broadband  n_children_computer_no_internet  \
0                                978                               66   
1                                383                                0   
2                                498                               23   
3                                619                               18   
4                                434                               24   

   n_children_no_computer  
0                       0  
1                       0  
2                       0  
3                       8  
4                      17

In [13]:
data2 = pd.read_csv(fccfile)
data2 = data2.rename(columns={'tract':'GEOID'})
data2['GEOID'] = data2['GEOID'].apply(lambda x: str(x))
data2.drop(columns=['state'], inplace=True)
data2.head()

GEOID       max_dn       max_up      dn10     dn100     dn250  \
0  17001000100   931.548571   915.160000  2.268571  1.502857  1.502857   
1  17001000201   872.587302   796.936508  2.761905  1.603175  1.603175   
2  17001000202  1000.000000  1000.000000  3.944444  2.018519  2.018519   
3  17001000400   519.804124   198.982021  1.757732  0.680412  0.680412   
4  17001000500   982.120690   916.275862  3.775862  1.793103  1.793103   

   fiber_100u  
0    0.920000  
1    0.809524  
2    1.018519  
3    0.190722  
4    0.913793

In [14]:
data3 = pd.read_csv(ooklafile)
wm = lambda x: np.average(x, weights=data3.loc[x.index, 'tests'])
data3a = data3.groupby(['GEOID','type'], as_index=False).agg(
            d_kbps_wm = ('avg_d_kbps', wm),
            u_kbps_wm = ('avg_u_kbps', wm)
)
data3b = data3.groupby('GEOID', as_index=False).agg({'tests': 'sum', 'devices': 'sum'})
fixed = data3a[data3a.type == 'fixed']
mobile = data3a[data3a.type == 'mobile']
data3 = mobile.merge(fixed, on='GEOID')
data4 = data3.merge(data3b, on="GEOID")
data4['GEOID'] = data4['GEOID'].apply(lambda x: str(x))
data4.head()

GEOID  type_x   d_kbps_wm_x   u_kbps_wm_x type_y    d_kbps_wm_y  \
0  17001000100  mobile  42029.527273   6680.036364  fixed  182120.250000   
1  17001000201  mobile  54537.523810  18398.714286  fixed  136686.000000   
2  17001000202  mobile  52001.555556  16057.444444  fixed   94081.971429   
3  17001000400  mobile  20826.500000   6379.625000  fixed  114393.125000   
4  17001000500  mobile  29641.000000   5213.900000  fixed   91572.125000   

    u_kbps_wm_y  tests  devices  
0  17408.857143     83       40  
1  11395.320000     46       22  
2   9165.028571     44       22  
3   9905.750000     48       25  
4  12561.093750     42       24

In [15]:
df_1 = shpfile.merge(data, how="left", on="GEOID")
df_2 = df_1.merge(data2, how="left", on="GEOID")
df_final = df_2.merge(data4, how="left", on="GEOID")
df_final.head()

STATEFP COUNTYFP TRACTCE              AFFGEOID        GEOID  NAME LSAD  \
0      17      001  000100  1400000US17001000100  17001000100     1   CT   
1      17      001  000201  1400000US17001000201  17001000201  2.01   CT   
2      17      001  000202  1400000US17001000202  17001000202  2.02   CT   
3      17      001  000400  1400000US17001000400  17001000400     4   CT   
4      17      001  000500  1400000US17001000500  17001000500     5   CT   

        ALAND    AWATER                                           geometry  \
0  12937185.0   22042.0  POLYGON ((-91.37766 39.94160, -91.37759 39.946...   
1   3206335.0       0.0  POLYGON ((-91.39646 39.95621, -91.39631 39.965...   
2   1514956.0       0.0  POLYGON ((-91.39370 39.94678, -91.37759 39.946...   
3   4458010.0  211712.0  POLYGON ((-91.42005 39.95081, -91.41917 39.951...   
4   1262517.0       0.0  POLYGON ((-91.40340 39.95048, -91.39655 39.950...   

   f_broadband  f_computer   f_ba  f_black  f_hispanic  log_mhi      mhi  \
0        0.813       0.861  0.246    0.068       0.000    10.78  48088.0   
1        0.879       0.940  0.266    0.078       0.016    10.71  44855.0   
2        0.863       0.933  0.221    0.080       0.008    10.83  50375.0   
3        0.691       0.780  0.094    0.182       0.006    10.31  30164.0   
4        0.665       0.787  0.113    0.133       0.035    10.62  41008.0   

   population  households  broadband  computer  black  hispanic    mhi.1  \
0      4761.0      2202.0       81.3      86.1    6.8       0.0  48088.0   
1      1985.0       889.0       87.9      94.0    7.8       1.6  44855.0   
2      2376.0       970.0       86.3      93.3    8.0       0.8  50375.0   
3      3422.0      1316.0       69.1      78.0   18.2       0.6  30164.0   
4      2175.0       813.0       66.5      78.7   13.3       3.5  41008.0   

     ba  den_computers  n_computer  n_broadband  den_black  n_black  \
0  24.6         2202.0      1896.0       1791.0     4761.0    324.0   
1  26.6          889.0       836.0        781.0     1985.0    155.0   
2  22.1          970.0       905.0        837.0     2376.0    191.0   
3   9.4         1316.0      1027.0        909.0     3422.0    624.0   
4  11.3          813.0       640.0        541.0     2175.0    290.0   

   den_hispanic  n_hispanic  den_ba   n_ba  nhh_computer  nhh_broadband  \
0        4761.0         0.0  3381.0  831.0          1896           1791   
1        1985.0        32.0  1400.0  373.0           836            781   
2        2376.0        19.0  1527.0  338.0           905            837   
3        3422.0        20.0  2536.0  238.0          1027            909   
4        2175.0        77.0  1455.0  165.0           640            541   

   nhh_computer_any_internet  nhh_computer_and_dialup  \
0                       1896                        0   
1                        836                        5   
2                        905                        7   
3                       1027                        9   
4                        640                        0   

   nhh_computer_and_broadband  nhh_computer_no_internet  nhh_no_computer  \
0                        1791                       105              306   
1                         781                        50               53   
2                         829                        69               65   
3                         909                       109              289   
4                         536                       104              173   

   den_age  n_children  n_children_computer  n_children_computer_and_dialup  \
0     4741        1044                 1044                               0   
1     1961         383                  383                               0   
2     2336         521                  521                               0   
3     3147         645                  637                               0   
4     1878         475                  458                               0   

   n_chil

In [82]:
df_final.to_csv('il_tot.csv')

# Formula to create dataset

In [3]:
shapefile = "/content/drive/MyDrive/Datasets/Broadband/IL/il_spdf/il_spdf.shp"
acsfile = '/content/drive/MyDrive/Datasets/Broadband/IL/acs_2019_IL.csv'
fccfile = '/content/drive/MyDrive/Datasets/Broadband/IL/fcc_477_census_tract_IL.csv'
ooklafile = '/content/drive/MyDrive/Datasets/Broadband/IL/ookla_combined_il.csv'

In [16]:
def prepare_dataset(shapefile, acsfile, fccfile, ooklafile):
    shpfile = gpd.read_file(shapefile)
    shpfile = shpfile.sort_values(by=['GEOID'])
    data = pd.read_csv(acsfile)
    data = data.rename(columns={'geoid':'GEOID'})
    data['GEOID'] = data['GEOID'].apply(lambda x: str(x))
    data.drop(columns=['state', 'county', 'tract', 'state_lkp'], inplace=True)
    data2 = pd.read_csv(fccfile)
    data2 = data2.rename(columns={'tract':'GEOID'})
    data2['GEOID'] = data2['GEOID'].apply(lambda x: str(x))
    data2.drop(columns=['state'], inplace=True)
    data3 = pd.read_csv(ooklafile)
    wm = lambda x: np.average(x, weights=data3.loc[x.index, 'tests'])
    data3a = data3.groupby(['GEOID','type'], as_index=False).agg(
                d_kbps_wm = ('avg_d_kbps', wm),
                u_kbps_wm = ('avg_u_kbps', wm)
    )
    data3b = data3.groupby('GEOID', as_index=False).agg({'tests': 'sum', 'devices': 'sum'})
    fixed = data3a[data3a.type == 'fixed']
    mobile = data3a[data3a.type == 'mobile']
    data3 = mobile.merge(fixed, on='GEOID')
    data4 = data3.merge(data3b, on="GEOID")
    data4['GEOID'] = data4['GEOID'].apply(lambda x: str(x))
    df_1 = shpfile.merge(data, how="left", on="GEOID")
    df_2 = df_1.merge(data2, how="left", on="GEOID")
    df_final = df_2.merge(data4, how="left", on="GEOID")
    return df_final


In [17]:
il_df = prepare_dataset(shapefile, acsfile, fccfile, ooklafile)
il_df

STATEFP COUNTYFP TRACTCE              AFFGEOID        GEOID    NAME LSAD  \
0         17      001  000100  1400000US17001000100  17001000100       1   CT   
1         17      001  000201  1400000US17001000201  17001000201    2.01   CT   
2         17      001  000202  1400000US17001000202  17001000202    2.02   CT   
3         17      001  000400  1400000US17001000400  17001000400       4   CT   
4         17      001  000500  1400000US17001000500  17001000500       5   CT   
...      ...      ...     ...                   ...          ...     ...  ...   
3116      17      203  030501  1400000US17203030501  17203030501  305.01   CT   
3117      17      203  030502  1400000US17203030502  17203030502  305.02   CT   
3118      17      203  030601  1400000US17203030601  17203030601  306.01   CT   
3119      17      203  030602  1400000US17203030602  17203030602  306.02   CT   
3120      17      203  030700  1400000US17203030700  17203030700     307   CT   

            ALAND      AWATER  \
0      12937185.0     22042.0   
1       3206335.0         0.0   
2       1514956.0         0.0   
3       4458010.0    211712.0   
4       1262517.0         0.0   
...           ...         ...   
3116   44516795.0    111544.0   
3117   35324557.0  22436047.0   
3118  107361660.0    157898.0   
3119  191120324.0   1123588.0   
3120  235145636.0    120371.0   

                                               geometry  f_broadband  \
0     POLYGON ((-91.37766 39.94160, -91.37759 39.946...        0.813   
1     POLYGON ((-91.39646 39.95621, -91.39631 39.965...        0.879   
2     POLYGON ((-91.39370 39.94678, -91.37759 39.946...        0.863   
3     POLYGON ((-91.42005 39.95081, -91.41917 39.951...        0.691   
4     POLYGON ((-91.40340 39.95048, -91.39655 39.950...        0.665   
...                                                 ...          ...   
3116  POLYGON ((-89.51271 40.76214, -89.51124 40.769...        0.886   
3117  POLYGON ((-89.56252 40.79285, -89.56239 40.793...        0.833   
3118  POLYGON ((-89.33009 40.74826, -89.31363 40.748...        0.793   
3119  POLYGON ((-89.32869 40.66078, -89.31987 40.660...        0.816   
3120  POLYGON ((-89.17682 40.75731, -89.15773 40.757...        0.754   

      f_computer   f_ba  f_black  f_hispanic  log_mhi       mhi  population  \
0          0.861  0.246    0.068       0.000    10.78   48088.0      4761.0   
1          0.940  0.266    0.078       0.016    10.71   44855.0      1985.0   
2          0.933  0.221    0.080       0.008    10.83   50375.0      2376.0   
3          0.780  0.094    0.182       0.006    10.31   30164.0      3422.0   
4          0.787  0.113    0.133       0.035    10.62   41008.0      2175.0   
...          ...    ...      ...         ...      ...       ...         ...   
3116       0.942  0.526    0.014       0.017    11.52  100875.0      7795.0   
3117       0.906  0.168    0.002       0.015    11.02   61293.0      2500.0   
3118       0.876  0.384    0.020       0.034    11.00   59929.0      6620.0   
3119       0.922  0.327    0.013       0.005    11.40   89650.0      3437.0   
3120       0.868  0.237    0.025       0.005    11.20   73421.0      4226.0   

      households  broadband  computer  black  hispanic     mhi.1    ba  \
0         2202.0       81.3      86.1    6.8       0.0   48088.0  24.6   
1          889.0       87.9      94.0    7.8       1.6   44855.0  26.6   
2          970.0       86.3      93.3    8.0       0.8   50375.0  22.1   
3         1316.0       69.1      78.0   18.2       0.6   30164.0   9.4   
4          813.0       66.5      78.7   13.3       3.5   41008.0  11.3   
...          ...        ...       ...    ...       ...       ...   ...   
3116      2758.0       88.6      94.2    1.4       1.7  100875.0  52.6   
3117      1084.0       83.3      90.6    0.2       1.5   61293.0  16.8   
3118      2371.0       79.3      87.6    2.0       3.4   59929.0  38.4   
3119      1261.0       81.6      92.2    1.3       0.5   89650.0  32.7   
3120   